src/app/clusters/device-energy-management-server/device-energy-management-server.cpp

```diff
@@ -0,0 +1,517 @@
+{
+    Status status                                               = Status::Success;
+    uint32_t earliestStartTime                                  = 0;
+    uint32_t latestEndTime                                      = 0;
+    DataModel::Nullable<Structs::ForecastStruct::Type> forecast = mDelegate.GetForecast();
+
+    uint32_t duration;
+    uint32_t requestedStartTime = commandData.requestedStartTime;
+
+    if (forecast.IsNull())
+    {
+        ChipLogError(Zcl, "DEM: %s - Forecast is Null", __FUNCTION__);
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::Failure);
+        return;
+    }
+
+    if (ESAStateEnum::kUserOptOut == mDelegate.GetESAState())
+    {
+        ChipLogError(Zcl, "DEM: %s - ESAState = kUserOptOut", __FUNCTION__);
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::Failure);
+        return;
+    }
+
+    /* If the RequestedStartTime value resulted in a time shift which is
+     * outside the time constraints of EarliestStartTime and
+     * LatestEndTime, then the command SHALL be rejected with CONSTRAINT_ERROR;
+     * otherwise the command SHALL be rejected with FAILURE
+     */
+    /* earliestStartTime is optional based on the FA (ForecastAdjust) feature AND can be nullable */
+    if (forecast.Value().earliestStartTime.HasValue() || forecast.Value().latestEndTime.HasValue())
+    {
+        /* These Should not be NULL since this command requires FA feature and these are mandatory for that */
+        ChipLogError(Zcl, "DEM: %s - EarliestStartTime / LatestEndTime not valid", __FUNCTION__);
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::Failure);
+        return;
+    }
+
+    chip::System::Clock::Milliseconds64 cTMs;
+    auto chipError = chip::System::SystemClock().GetClock_RealTimeMS(cTMs);
+    if (chipError != CHIP_NO_ERROR)
+    {
+        ChipLogError(Zcl, "DEM: Unable to get current time - error=%d (%s)]", chipError.AsInteger(), chipError.AsString());
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::Failure);
+        return;
+    }
+    uint32_t epochNow = std::chrono::duration_cast<chip::System::Clock::Seconds32>(cTMs).count();
+
+    if (forecast.Value().earliestStartTime.Value().IsNull())
+    {
+        /* Null means - We can start immediately */
+        earliestStartTime = epochNow; /* NOW */
+    }
+    else
+    {
+        earliestStartTime = forecast.Value().earliestStartTime.Value().Value();
+    }
+    /* Latest End Time is optional & cannot be null - unlike earliestStartTime! */
+    latestEndTime = forecast.Value().latestEndTime.Value();
+
+    duration = forecast.Value().endTime - forecast.Value().startTime; // the current entire forecast duration
+
+    if ((requestedStartTime < earliestStartTime) || ((requestedStartTime + duration) > latestEndTime))
+    {
+        ChipLogError(Zcl, "DEM: %s - Bad requestedStartTime %d.", __FUNCTION__, requestedStartTime);
```

Comment:
May be useful to indicate the offshoot in the log ?!

Reply:
Done